File: mg3d.py
#!/usr/bin/env python3
"""咪咕音乐 .mg3d 解密（算法移植自 ioococ/music-unlock 的 mg3d.ts）。
原理：减法密码——文件中藏着一段 0x20 字节的大写十六进制字符串作密钥，
密文逐字节减去密钥循环。用 RIFF/WAVE 头结构反推验证密钥，无需外部凭据。"""
import os

SEG = 0x20


def _is_upper_hex(ch):
    return (0x30 <= ch <= 0x39) or (0x41 <= ch <= 0x46)


def _is_printable(ch):
    return 0x20 <= ch <= 0x7E


def _sub(data, key):
    return bytes(b - key[i % SEG] & 0xFF for i, b in enumerate(data))


def _valid_header(h):
    if h[:4] != b"RIFF" or h[8:16] != b"WAVEfmt ":
        return False
    fmt_size = int.from_bytes(h[0x10:0x14], "little")
    if fmt_size not in (16, 18, 40):
        return False
    off1 = 0x14 + fmt_size
    if not all(_is_printable(c) for c in h[off1:off1 + 4]):
        return False
    off2 = off1 + 8 + int.from_bytes(h[off1 + 4:off1 + 8], "little")
    if off2 <= len(h) and not all(_is_printable(c) for c in h[off2:off2 + 4]):
        return False
    return True


def mg3d_decrypt(path, outdir):
    """返回 (成功与否, 输出路径或失败原因)。"""
    with open(path, "rb") as f:
        buf = f.read()
    header = buf[:0x100]
    key = None
    for off in range(SEG, SEG * 20, SEG):
        cand = buf[off:off + SEG]
        if len(cand) < SEG or not all(_is_upper_hex(c) for c in cand):
            continue
        if _valid_header(_sub(header, cand)):
            key = cand
            break
    if key is None:
        return False, "未找到咪咕密钥（非 mg3d 格式？）"
    data = _sub(buf, key)
    os.makedirs(outdir, exist_ok=True)
    out = os.path.join(outdir, os.path.splitext(os.path.basename(path))[0] + ".wav")
    with open(out, "wb") as f:
        f.write(data)
    return True, out

File: crates/core/src/mg3d.rs
//! 咪咕 .mg3d：减法密码，对照 mg3d.py。

use std::fs;
use std::path::{Path, PathBuf};

const SEG: usize = 0x20;

fn is_upper_hex(ch: u8) -> bool {
    ch.is_ascii_digit() || (b'A'..=b'F').contains(&ch)
}

fn is_printable(ch: u8) -> bool {
    (0x20..=0x7E).contains(&ch)
}

fn sub(data: &[u8], key: &[u8]) -> Vec<u8> {
    data.iter()
        .enumerate()
        .map(|(i, b)| b.wrapping_sub(key[i % SEG]))
        .collect()
}

fn valid_header(h: &[u8]) -> bool {
    if h.len() < 0x18 || &h[..4] != b"RIFF" || &h[8..16] != b"WAVEfmt " {
        return false;
    }
    let fmt_size = u32::from_le_bytes(h[0x10..0x14].try_into().unwrap()) as usize;
    if !matches!(fmt_size, 16 | 18 | 40) {
        return false;
    }
    let off1 = 0x14 + fmt_size;
    if off1 + 8 > h.len() || !h[off1..off1 + 4].iter().all(|&c| is_printable(c)) {
        return false;
    }
    let chunk = u32::from_le_bytes(h[off1 + 4..off1 + 8].try_into().unwrap()) as usize;
    let off2 = off1 + 8 + chunk;
    if off2 + 4 <= h.len() && !h[off2..off2 + 4].iter().all(|&c| is_printable(c)) {
        return false;
    }
    true
}

/// 成功返回输出 wav 路径。
pub fn decrypt(path: &Path, outdir: &Path) -> Result<PathBuf, String> {
    let buf = fs::read(path).map_err(|e| e.to_string())?;
    let header_end = buf.len().min(0x100);
    let mut key = None;
    let mut off = SEG;
    while off < SEG * 20 {
        let end = off + SEG;
        if end > buf.len() {
            break;
        }
        let cand = &buf[off..end];
        if cand.iter().all(|&c| is_upper_hex(c)) {
            let header = sub(&buf[..header_end], cand);
            if valid_header(&header) {
                key = Some(cand.to_vec());
                break;
            }
        }
        off += SEG;
    }
    let key = key.ok_or_else(|| "未找到咪咕密钥（非 mg3d 格式？）".to_string())?;
    let data = sub(&buf, &key);
    fs::create_dir_all(outdir).map_err(|e| e.to_string())?;
    let stem = path.file_stem().unwrap_or_default().to_string_lossy();
    let out = outdir.join(format!("{stem}.wav"));
    fs::write(&out, data).map_err(|e| e.to_string())?;
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    fn tiny_wav() -> Vec<u8> {
        let data = vec![0u8; 160];
        let mut out = Vec::new();
        out.extend(b"RIFF");
        out.extend(&(36u32 + data.len() as u32).to_le_bytes());
        out.extend(b"WAVEfmt ");
        out.extend(&16u32.to_le_bytes());
        out.extend(&1u16.to_le_bytes());
        out.extend(&1u16.to_le_bytes());
        out.extend(&8000u32.to_le_bytes());
        out.extend(&16000u32.to_le_bytes());
        out.extend(&2u16.to_le_bytes());
        out.extend(&16u16.to_le_bytes());
        out.extend(b"data");
        out.extend(&(data.len() as u32).to_le_bytes());
        out.extend(data);
        out
    }

    #[test]
    fn synthetic_wav_roundtrip() {
        let wav = tiny_wav();
        assert_eq!(&wav[0x40..0x60], &[0u8; 32]);
        let key = b"0123456789ABCDEF0123456789ABCDEF";
        let enc: Vec<u8> = wav
            .iter()
            .enumerate()
            .map(|(i, b)| b.wrapping_add(key[i % 32]))
            .collect();
        let dir = tempdir().unwrap();
        let src = dir.path().join("song.mg3d");
        fs::write(&src, enc).unwrap();
        let out = decrypt(&src, dir.path()).unwrap();
        assert_eq!(fs::read(out).unwrap(), wav);
    }

    #[test]
    fn garbage_is_rejected() {
        let dir = tempdir().unwrap();
        let src = dir.path().join("nope.mg3d");
        fs::write(&src, b"not a mg3d file".repeat(20)).unwrap();
        let err = decrypt(&src, dir.path()).unwrap_err();
        assert!(err.contains("密钥"));
    }
}
